File: packages/vc/src/canon.test.ts
import { readFileSync } from 'node:fs'
import { afterEach, describe, expect, it, vi } from 'vitest'
import { canonicalize } from './canon.js'
import { CLAIMS_V1, CREDENTIALS_V2, staticLoader, UnknownContextError } from './context/loader.js'

const vector = (name: string) => readFileSync(new URL(`../../../vectors/canon/${name}`, import.meta.url), 'utf8')
const minimal = () => JSON.parse(vector('minimal.json')) as Record<string, unknown>

afterEach(() => vi.restoreAllMocks())

describe('staticLoader', () => {
  it('serves the two bundled contexts', async () => {
    const load = staticLoader()
    expect((await load(CREDENTIALS_V2)).document).toHaveProperty('@context')
    expect((await load(CLAIMS_V1)).document).toHaveProperty('@context')
  })

  it('refuses anything else', async () => {
    await expect(staticLoader()('https://example.com/ctx')).rejects.toBeInstanceOf(UnknownContextError)
    await expect(staticLoader()('constructor')).rejects.toBeInstanceOf(UnknownContextError)
  })

  it('does not let extra documents replace a bundled context', async () => {
    const load = staticLoader({ [CLAIMS_V1]: { '@context': {} } })
    expect((await load(CLAIMS_V1)).document).toEqual((await staticLoader()(CLAIMS_V1)).document)
  })
})

describe('canonicalize', () => {
  it('reproduces the committed N-Quads', async () => {
    expect(await canonicalize(minimal())).toBe(vector('minimal.nq'))
  })

  it('does not depend on key order', async () => {
    const doc = minimal()
    const reversed = Object.fromEntries(Object.entries(doc).reverse())
    reversed.credentialSubject = Object.fromEntries(Object.entries(doc.credentialSubject as object).reverse())
    expect(await canonicalize(reversed)).toBe(vector('minimal.nq'))
  })

  it('rejects a property no context defines', async () => {
    const doc = minimal()
    ;(doc.credentialSubject as Record<string, unknown>).balance = '1000'
    await expect(canonicalize(doc)).rejects.toThrow()
  })

  it('rejects a type no context defines', async () => {
    const doc = minimal()
    doc.type = ['VerifiableCredential', 'NotDefinedAnywhere']
    await expect(canonicalize(doc)).rejects.toThrow()
  })

  it('rejects a relative id', async () => {
    const doc = minimal()
    doc.id = 'not-an-iri'
    await expect(canonicalize(doc)).rejects.toThrow()
  })

  it('rejects an unknown context without fetching it', async () => {
    const fetchSpy = vi.spyOn(globalThis, 'fetch')
    const doc = minimal()
    ;(doc['@context'] as string[]).push('https://example.com/ctx')
    await expect(canonicalize(doc)).rejects.toThrow(/example\.com/)
    expect(fetchSpy).not.toHaveBeenCalled()
  })
})

File: vc-go/canon/canon_test.go
package canon

import (
	"errors"
	"os"
	"path/filepath"
	"testing"

	"github.com/vemphy/vc/vc-go/internal/vectors"
	"github.com/vemphy/vc/vc-go/vcctx"
)

func minimal(t *testing.T) map[string]any {
	var doc map[string]any
	vectors.Load(t, "canon/minimal.json", &doc)
	return doc
}

func TestReproducesCommittedNQuads(t *testing.T) {
	want, err := os.ReadFile(filepath.Join(vectors.Dir(t), "canon", "minimal.nq"))
	if err != nil {
		t.Fatal(err)
	}
	got, err := Canonicalize(minimal(t), nil)
	if err != nil {
		t.Fatal(err)
	}
	if got != string(want) {
		t.Errorf("N-Quads differ from the TypeScript output\n--- got\n%s--- want\n%s", got, want)
	}
}

func TestRejectsUndefinedProperty(t *testing.T) {
	doc := minimal(t)
	doc["credentialSubject"].(map[string]any)["balance"] = "1000"
	if _, err := Canonicalize(doc, nil); err == nil {
		t.Error("expected an error for a property no context defines")
	}
}

func TestRejectsUnknownContext(t *testing.T) {
	doc := minimal(t)
	doc["@context"] = append(doc["@context"].([]any), "https://example.com/ctx")
	_, err := Canonicalize(doc, nil)
	if !errors.Is(err, vcctx.ErrUnknownContext) {
		t.Errorf("got %v, want ErrUnknownContext", err)
	}
}

func TestRejectsUndefinedType(t *testing.T) {
	doc := minimal(t)
	doc["type"] = []any{"VerifiableCredential", "NotDefinedAnywhere"}
	if _, err := Canonicalize(doc, nil); err == nil {
		t.Error("expected an error for a type no context defines")
	}
}

func TestRejectsRelativeID(t *testing.T) {
	doc := minimal(t)
	doc["id"] = "not-an-iri"
	if _, err := Canonicalize(doc, nil); err == nil {
		t.Error("expected an error for a relative id")
	}
}
